compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/FunctionalReference.kt

```diff
@@ -0,0 +1,55 @@
+package com.squareup.anvil.compiler.internal.reference
+
+import com.squareup.anvil.annotations.ExperimentalAnvilApi
+import com.squareup.anvil.compiler.api.AnvilCompilationException
+import com.squareup.anvil.compiler.internal.reference.FunctionalReference.Descriptor
+import com.squareup.anvil.compiler.internal.reference.FunctionalReference.Psi
+import org.jetbrains.kotlin.descriptors.FunctionDescriptor
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.KtFunction
+
+@ExperimentalAnvilApi
+public sealed interface FunctionalReference {
```

Comment:
`FunctionalReference` seems a bit of an odd name, vs maybe doing `FunctionReference` with subtypes of `TopLevelFunctionReference` and `RegularFunctionReference`? I admit at a glance I'm not sure I followed why they need to be separate implementations

Reply:
We'd need to rename the current class, which would also be a significant breaking change for consumers and custom code generators. We don't give any API stability guarantees, but nonetheless I wanted to avoid this. Initially I considered:
```
FunctionReference
-> MemberFunctionReference
-> TopLevelFunctionReference
```

> I admit at a glance I'm not sure I followed why they need to be separate implementations

Answer is here: https://github.com/square/anvil/pull/675#discussion_r1067292750

Reply:
Gotcha. I admit the name doesn't feel great but I don't have a better alternative to suggest!